guess-sig.1
.ds = Ns = Ns
.Dd October 23, 2018
.Dt GUESS-SIG 1
.Os
.Sh NAME
.Nm guess-sig
.Nd display matching byte sequences in file headers
.Sh SYNOPSIS
.Nm
.Op Fl c Ar columns
.Op Fl b Ar bytes
.Op Fl f Ar format
.Op Fl -dot-hack
.Op Fl -hex
.Op Fl -json
.Op Fl -regex
.Ar files
.Nm
.Op Fl v Fl -version
.Sh DESCRIPTION
.Nm
searches the given input
.Ar files
for common byte sequences, emitting those with identical values and offsets.
Its main purpose is to help identify the signatures of undocumented or proprietary file formats.
.
.Pp
The following options are available:
.
.Bl -tag -width indent
.It Fl c Ar num , Fl -cols \*= Ar num , Fl -columns \*= Ar number
Line-wrap output to fit
.Ar num
columns, used only when
.Fl -format
is
.Ql hex .
The default is
.Em 16 .
.
.It Fl b Ar num , Fl n Ar num , Fl -bytes \*= Ar number
Limit scanning to
.Ar num
bytes per file.
The default limit is
.Em 512
bytes.
.
.It Fl d , Fl -dot-hack
Generate regexp output compatible with older JavaScript engines.
Implies
.Fl -format \*= regexp .
.
.It Fl f Ar type , Fl -format \*= Ar type
Specify the output format for guessed file signatures.
Supported
.Ar types
are
.Ql hex ,
.Ql json ,
and
.Ql regex .
The default is
.Ql hex .
Each supported
.Ar type
has a corresponding switch of the same name, described below.
.
.It Fl h , Fl -hex
Display matching bytes as hexadecimal, and non-matching bytes as
.Ql __ .
This is the default format.
.
.It Fl j , Fl -json
Emit a JSON array where matching bytes are numbers, and non-matching bytes are
.Ql null .
Analogous to
.Dq Fl -format \*= Ar json .
.
.It Fl r , Fl -regex , Fl -regexp
Output an ECMAScript RegExp literal for matching the guessed file signature.
Analogous to
.Dq Fl -format \*= Ar regex .
.Pp
Note that generated patterns use the \(lqdotall\(rq modifier,
which is not supported by older JavaScript engines.
If portability is a concern, include the
.Fl -dot-hack
switch.
.
.It Fl v , Fl -version
Print the program's version string, then exit.
.El
.
.Sh EXIT STATUS
Exits with an error code of
.Sy 1
if no files were given, or no meaningful data was found.
.
.Sh EXAMPLES
Display matching bytes as hex values:
.
.Bd -literal -offset 6n
$ guess-sig ~/Pictures/*.jpg

FF D8 FF E0 00 10 4A 46 49 46 00 01 01 __ 00 __
00 __ 00 00 FF __ 00 __ __ __ __ __ __ __ __ __
.Ed
.Pp
Generate machine-readable JSON data:
.Bd -literal -offset 6n
$ guess-sig -j ~/Pictures/*.jpg

[255,216,255,224,0,16,74, \.\.\. 70,73,70,0,1,1,null,0,null]
.Ed
.
.Pp
Generate a RegExp literal to match the signature in JavaScript:
.Bd -literal -offset 6n
$ guess-sig -r ~/Pictures/*.jpg

/^\e0\exFFKA\.{4}nsrP#Inf\.{48}Prot\.{2}\e0\e0#Filx\ex01/s
.Ed
.
.Sh DIAGNOSTICS
Warnings are emitted to standard error if passed an empty or non-existent file,
or if entities aren't a regular file.
.
.Sh SEE ALSO
.Xr file 1 ,
.Xr xxd 1
.
.Sh AUTHORS
Copyright \(co 2018
.An "John Gardner"
.Aq Mt gardnerjohng@gmail.com .


guess-sig.js
#!/usr/bin/env node
"use strict";

const fs      = require("fs");
const getOpts = require("get-options");
const headers = [];

// Parse options
const {options, argv} = getOpts(process.argv.slice(2), {
	"-c, --cols, --columns": "[number=\\d+]",
	"-b, --bytes, -n":       "[number=\\d+]",
	"-f, --format":          "[type]",
	"-d, --dot-hack":        "",
	"-h, --hex":             "",
	"-j, --json":            "",
	"-r, --regex, --regexp": "",
	"-v, --version":         "",
	"--help":                "",
});
let format     = options.format   || "hex";
const columns  = +options.columns || 16;
const numBytes = +options.bytes   || 512;
const dotHack  = !!options.dotHack;
if(options.json)  format = "json";
if(options.regex) format = "regex";
if(options.hex)   format = "hex";

// Print a usage summary and exit
if(options.help){
	showUsage();
	process.exit(0);
}

// Print version string and exit
if(options.version){
	console.log(require("./package.json").version);
	process.exit(0);
}

// Bail if no arguments were given
if(!argv.length){
	showUsage();
	process.exit(1);
}

// Collect header samples from each file named on command-line
for(let file of argv){
	
	// Ignore missing files
	if(!fs.existsSync(file)){
		console.warn(`Skipping non-existent file: ${file}`);
		continue;
	}
	
	let stats = fs.lstatSync(file);
	
	// Resolve symlinks
	if(stats.isSymbolicLink()){
		file = fs.readlinkSync(file);
		stats = fs.lstatSync(file);
	}
	
	// Not a regular file
	if(!stats.isFile(file)){
		console.warn(`Not a regular file: ${file}`);
		continue;
	}
	
	// Ignore empty files
	if(!stats.size){
		console.warn(`Skipping empty file: ${file}`);
		continue;
	}

	const fd = fs.openSync(file, "r");
	const buffer = Buffer.alloc(Math.min(numBytes, stats.size));
	fs.readSync(fd, buffer, 0, numBytes);
	fs.closeSync(fd);
	headers.push(buffer);
}

// Nothing to scan? Bail
const numFiles = headers.length;
if(!numFiles){
	console.error("Nothing to compare");
	process.exit(1);
}

let table = new Array(numBytes);
for(let i = 0; i < numBytes; ++i)
	table[i] = new Array(numFiles);

for(let i = 0; i < numFiles; ++i)
	for(let j = 0; j < numBytes; ++j)
		table[j][i] = headers[i][j];


table = table.map(column => isHomogenous(column) ? column[0] : null);

switch(format){
	case "json":
		console.log(JSON.stringify(table));
		break;
	
	// Print an ECMAScript RegExp literal for matching the guessed signature
	case "regex": {
		const patterns = [];
		let prevByte = false;
		let repeated = 0;
		
		for(let i = 0; i < numBytes; ++i){
			const byte = table[i];
			if(prevByte === byte)
				++repeated;
			else{
				if(false !== prevByte)
					patterns.push(regexify(prevByte, repeated, dotHack));
				prevByte = byte;
				repeated = 0;
			}
		}
		null !== prevByte && patterns.push(regexify(prevByte, repeated));
		process.stdout.write(`/^${patterns.join("")}/${dotHack ? "" : "s"}`);
		process.stdout.isTTY && process.stdout.write("\n");
		break;
	}
	
	case "hex":
		for(let i = 0; i < numBytes; ++i){
			i && process.stdout.write(i % columns ? " " : "\n");
			process.stdout.write(null === table[i] ? "__" : hex(table[i]));
		}
		process.stdout.write("\n");
		break;
}


/**
 * Return true if every element of an array is identical.
 * @param {Array}
 * @return {Boolean}
 */
function isHomogenous(array){
	const {length} = array;
	for(let i = 1; i < length; ++i)
		if(array[i] !== array[0])
			return false;
	return true;
}


/**
 * Convert an integer to uppercased hexadecimal.
 * @example hex(12) => "0C"
 * @param {Number} byte
 * @return {String}
 */
function hex(byte){
	return (byte < 0x10 ? "0" : "") + byte.toString(16).toUpperCase();
}


/**
 * Convert a byte (with possible repetition) to a regexp pattern.
 * @param {Number} byte
 * @param {Number} [repeatCount=0]
 * @param {Boolean} [dotHack=false]
 * @return {String}
 */
function regexify(byte, repeatCount = 0, dotHack = false){
	const quant = repeatCount > 0 ? `{${repeatCount + 1}}` : "";
	
	switch(byte){
		case 0x00: return `\\0${quant}`;  // Null byte
		case 0x09: return `\\t${quant}`;  // Tab
		case 0x0A: return `\\n${quant}`;  // Newline
		case 0x0B: return `\\v${quant}`;  // Vertical tab
		case 0x0C: return `\\f${quant}`;  // Form feed
		case 0x0D: return `\\r${quant}`;  // Carriage return
		
		// Metacharacters
		case 0x24: case 0x28: case 0x29: case 0x2A: case 0x2B:
		case 0x2E: case 0x3F: case 0x5B: case 0x5C: case 0x5D:
		case 0x5E: case 0x7B: case 0x7C: case 0x7D:
			return `\\${String.fromCharCode(byte)}${quant}`;
		
		// Match any byte
		case null:
			return dotHack ? `(?:.|[^\\0])${quant}` : `.${quant}`;
		
		default:
			// Printable ASCII range
			if(byte > 31 && byte < 127)
				return String.fromCharCode(byte) + quant;
			
			// Codepoint escape for unprintable or high-bit character
			return `\\x${hex(byte)}${quant}`;
	}
}


/**
 * Display a brief help summary.
 */
function showUsage(){
	// Bold, underline, off
	const [B, U, O] = process.stdout.isTTY
		? ["\x1B[1m", "\x1B[4m", "\x1B[0m"]
		: ["", "", ""];
	let usage = [
		`${B}Usage:${O} guess-sig`,
		`[-${B}dhjr${O}]`,
		`[-${B}c${O} ${U}cols${O}]`,
		`[-${B}b${O} ${U}bytes${O}]`,
		`[-${B}f${O} ${U}format${O}] ${U}...files${O}`,
	].join(" ");
	if("win32" !== process.platform)
		usage += "\n\nRun `man guess-sig' for full documentation.";
	console.log(usage);
}
